openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+#ifdef BENCHMARK_FAST_SWEEPING
+      timer.stop();
+#endif
+    }
+}// FastSweeping::sweep
+
+/// Private class of FastSweeping to quickly compute the extrema
+/// values of the active voxels in the leaf nodes. Several orders
+/// of magnitude faster than tools::extrema!
+template <typename GridT>
+struct FastSweeping<GridT>::MinMaxKernel
+{
+    using LeafMgr = tree::LeafManager<const TreeT>;
+    using LeafRange = typename LeafMgr::LeafRange;
+    MinMaxKernel() : mMin(std::numeric_limits<ValueT>::max()), mMax(-mMin) {}
+    MinMaxKernel(MinMaxKernel& other, tbb::split) : mMin(other.mMin), mMax(other.mMax) {}
+    math::MinMax<ValueT> run(const GridT &grid)
+    {
+        LeafMgr mgr(grid.tree());// super fast
+        tbb::parallel_reduce(mgr.leafRange(), *this);
+        return math::MinMax<ValueT>(mMin, mMax);
+    }
+    void operator()(const LeafRange& r)
+    {
+        for (auto leafIter = r.begin(); leafIter; ++leafIter) {
+            for (auto voxelIter = leafIter->beginValueOn(); voxelIter; ++voxelIter) {
+                const ValueT v = *voxelIter;
+                if (v < mMin) mMin = v;
+                if (v > mMax) mMax = v;
+            }
+        }
+    }
+    void join(const MinMaxKernel& other)
+    {
+        if (other.mMin < mMin) mMin = other.mMin;
+        if (other.mMax > mMax) mMax = other.mMax;
+    }
+    ValueT mMin, mMax;
+};//MinMaxKernel
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// Private class of FastSweeping to perform multi-threaded initialization
+template <typename GridT>
+struct FastSweeping<GridT>::DilateKernel
+{
+    using LeafRange = typename tree::LeafManager<TreeT>::LeafRange;
+    using BufferT = typename FastSweeping::CoordArrayT::ValueBuffer;//util::PagedArray<Coord, 10>::ValueBuffer;
+    using PoolT = tbb::enumerable_thread_specific<BufferT>;
+    DilateKernel(FastSweeping &parent)
+        : mParent(&parent), mPool(nullptr), mBackground(parent.mGrid1->background())
+    {
+    }
+    DilateKernel(const DilateKernel &parent) = default;// for tbb::parallel_for
+    DilateKernel& operator=(const DilateKernel&) = delete;
+
+    void run(int dilation, NearestNeighbors nn)
+    {
+#ifdef BENCHMARK_FAST_SWEEPING
+        util::CpuTimer timer("Construct LeafManager");
+#endif
+        tree::LeafManager<TreeT> mgr(mParent->mGrid1->tree());// super fast
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Changing background value");
+#endif
+        static const ValueT Unknown = std::numeric_limits<ValueT>::max();
+        changeLevelSetBackground(mgr, Unknown);//multi-threaded
+
+ #ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Dilating and updating mgr (parallel)");
+        //timer.restart("Dilating and updating mgr (serial)");
+#endif
+
+        const int delta = 5;
+        for (int i=0, d = dilation/delta; i<d; ++i) dilateActiveValues(mgr, delta, nn, IGNORE_TILES);
+        dilateActiveValues(mgr, dilation % delta, nn, IGNORE_TILES);
```

Comment:
Why have these multiple calls to `dilateActiveValues`? Is this the same as:
```
Why is this call to `dilateActiveValues` in a loop? Could you do:
```
const int iter = delta * (dilation/delta) + (dilation % delta);
dilateActiveValues(mgr, iter , nn, IGNORE_TILES);
```

Reply:
I believe I found this to be faster. Maybe that's no longer the case with Nick's changes?!

Reply:
To clarify previous discussion, we'vs decided to leave this as is and update it once #754 is accepted